### ISSUE TITLE
"paint" beatmaps

### ISSUE THREAD

**Issue by theonlygusti** [ppy/osu#6608]
Instead of using the current beatmap editor, allow someone to play the song on top of a blank beatmap, and when they "click" it puts a circle there, and when they click and drag it puts a slider there.

I think this could be a much more fun and easy way to create beatmaps, definitely in the beginning.

Maybe once all the sliders and circles are put using this method, their positions can be fine-tuned (and combos can be chosen etc.) in the traditional editor.

**Comment by TsavyPrince** [comment]
I support this.

**Comment by peppy** [comment]
This is called live mapping. I believe it is supported in stable and will also come to lazer eventually.

**Comment by enoslayd** [comment]
Will this be standard only?

**Comment by peppy** [comment]
Editor features will be present in all bundled rulesets.

**Comment by ViRiXDreamcore** [comment]
Has this been added to Lazer yet? Was trying to map a song and wanted to see if it made it in before I opened a request for it.

**Comment by D-Maxwell** [comment]
> Has this been added to Lazer yet? Was trying to map a song and wanted to see if it made it in before I opened a request for it.

Doesn't seem like it.

**Comment by nodingneu** [comment]
IMO this is more important than any other type of editor. If osu! only had one way to create maps, it should be live mapping.

**Comment by peppy** [comment]
You can basically do this now? Have you tried?

**Comment by snalgae** [comment]
I think this should be closed, live mapping is possible.

https://github.com/ppy/osu/assets/104220939/b878b15e-2e9f-4f73-8905-f3e33bae68c4

**Comment by Joehuu** [comment]
The title is kind of misleading. "Live mapping" in stable uses *keyboard bindings* (instead of mouse) to place circles (ref: https://osu.ppy.sh/wiki/en/Client/Beatmap_editor/Menu#compose). Stable also doesn't support slider placement, which OP wanted and can be exclusive to lazer.

TL;DR:
- use keyboard bindings to place circles, and when holding key and dragging, place a slider instead (i.e. automatically switch to slider mode drawing)
- probably also add the hold and drag behavior to mouse
- decide whether this should be a toggle or always default behavior

**Comment by bdach** [comment]
Tried to do a barebones implementation of this for taiko and hit a few walls.

```diff
diff --git a/osu.Game.Rulesets.Taiko/Edit/DrawableTaikoEditorRuleset.cs b/osu.Game.Rulesets.Taiko/Edit/DrawableTaikoEditorRuleset.cs
index 217bb8139c..f206130dd2 100644
--- a/osu.Game.Rulesets.Taiko/Edit/DrawableTaikoEditorRuleset.cs
+++ b/osu.Game.Rulesets.Taiko/Edit/DrawableTaikoEditorRuleset.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Collections.Generic;
+using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Game.Beatmaps;
 using osu.Game.Configuration;
@@ -20,6 +21,12 @@ public DrawableTaikoEditorRuleset(Ruleset ruleset, IBeatmap beatmap, IReadOnlyLi
         {
         }
 
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            KeyBindingInputManager.Add(new TaikoLiveMapper());
+        }
+
         protected override void LoadComplete()
         {
             base.LoadComplete();
diff --git a/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs b/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
index 174b278d89..04fcad4262 100644
--- a/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
+++ b/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
@@ -68,7 +68,7 @@ protected override void LoadComplete()
         private void regenerateAutoplay()
         {
             var autoplayMod = drawableRuleset.Mods.OfType<ModAutoplay>().Single();
-            drawableRuleset.SetReplayScore(autoplayMod.CreateScoreFromReplayData(drawableRuleset.Beatmap, drawableRuleset.Mods));
+            //drawableRuleset.SetReplayScore(autoplayMod.CreateScoreFromReplayData(drawableRuleset.Beatmap, drawableRuleset.Mods));
         }
 
         private void addHitObject(HitObject hitObject)
@@ -85,7 +85,7 @@ private void removeHitObject(HitObject hitObject)
 
         public override bool PropagatePositionalInputSubTree => false;
 
-        public override bool PropagateNonPositionalInputSubTree => false;
+        //public override bool PropagateNonPositionalInputSubTree => false;
 
         public PlayfieldAdjustmentContainer CreatePlayfieldAdjustmentContainer() => drawableRuleset.CreatePlayfieldAdjustmentContainer();
 
```

- To listen to user bindings the `TaikoLiveMapper` has to be under a `RulesetInputManager`. Instantiating one is currently not obvious, and attempting to reuse existing...
- requires _both_ autoplay replays (which power simulation of gameplay during playback) to be disabled _and_ for the `PropagateNonPositionalInput` override to be gone.
- Also there are keybinding priority conflicts; with this patch only the left default bindings work, the right ones are eaten by timeline.

The play here is probably to make instantiating the ruleset input manager easier. I can't imagine trying to jam the live mapper in the only current viable place is going to be nice or easy.
